dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java

```diff
@@ -23,72 +23,136 @@
 import java.util.regex.Pattern;
 
 import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
 
 import com.ibm.websphere.ras.Tr;
 import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.websphere.ras.annotation.Trivial;
 import com.ibm.ws.container.service.app.deploy.ModuleInfo;
+import com.ibm.ws.kernel.productinfo.ProductInfo;
 
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIAppConfigProvider;
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIAppConfigProvider.OpenAPIAppConfigListener;
+import io.openliberty.microprofile.openapi20.internal.services.ModuleSelectionConfig;
 import io.openliberty.microprofile.openapi20.internal.utils.Constants;
 import io.openliberty.microprofile.openapi20.internal.utils.MessageConstants;
 
 /**
  * Handles reading the merge include/exclude configuration properties and indicating whether a particular module should be included or excluded.
  */
-public class ModuleSelectionConfig {
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
+public class ModuleSelectionConfigImpl implements ModuleSelectionConfig, OpenAPIAppConfigProvider.OpenAPIAppConfigListener {
 
-    private static final TraceComponent tc = Tr.register(ModuleSelectionConfig.class);
+    private static final TraceComponent tc = Tr.register(ModuleSelectionConfigImpl.class);
 
+    private boolean hasBeenInit = false;
     private boolean isAll = false;
     private boolean isFirst = false;
+    private boolean serverxmlmode = false;
     private List<ModuleName> included;
     private List<ModuleName> excluded;
 
+    private OpenAPIAppConfigProvider configFromServerXML;
+
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC, unbind = "unbindAppConfigListener")
+    public void bindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        this.configFromServerXML = openAPIAppConfigProvider;
+        openAPIAppConfigProvider.registerAppConfigListener(this);
+    }
+
+    public void unbindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        openAPIAppConfigProvider.unregisterAppConfigListener(this);
+        this.configFromServerXML = null;
+    }
```

Comment:
Since this is a static reference, it might be easier to annotate the field and register as a listener in activate/deactivate.

Technically here we're registering as a listener before we're activated.